transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java

```diff
@@ -0,0 +1,654 @@
+import io.netty.channel.ChannelOutboundBuffer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.ConnectTimeoutException;
+import io.netty.channel.DefaultFileRegion;
+import io.netty.channel.EventLoop;
+import io.netty.channel.RecvByteBufAllocator;
+import io.netty.channel.socket.ChannelInputShutdownEvent;
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.StringUtil;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public abstract class AbstractEpollStreamChannel extends AbstractEpollChannel {
+
+    private static final String EXPECTED_TYPES =
+            " (expected: " + StringUtil.simpleClassName(ByteBuf.class) + ", " +
+                    StringUtil.simpleClassName(DefaultFileRegion.class) + ')';
+
+    private volatile boolean inputShutdown;
+    private volatile boolean outputShutdown;
+
+    protected AbstractEpollStreamChannel(Channel parent, int fd) {
+        super(parent, fd, Native.EPOLLIN, true);
+    }
+
+    protected AbstractEpollStreamChannel(int fd) {
+        super(fd, Native.EPOLLIN);
+    }
+
+    @Override
+    protected AbstractEpollUnsafe newUnsafe() {
+        return new EpollStreamUnsafe();
+    }
+
+    /**
+     * Write bytes form the given {@link ByteBuf} to the underlying {@link java.nio.channels.Channel}.
+     * @param buf           the {@link ByteBuf} from which the bytes should be written
+     */
+    private boolean writeBytes(ChannelOutboundBuffer in, ByteBuf buf) throws Exception {
+        int readableBytes = buf.readableBytes();
+        if (readableBytes == 0) {
+            in.remove();
+            return true;
+        }
+
+        if (buf.hasMemoryAddress() || buf.nioBufferCount() == 1) {
+            int writtenBytes = doWriteBytes(buf);
+            in.removeBytes(writtenBytes);
+            return writtenBytes == readableBytes;
+        } else {
+            ByteBuffer[] nioBuffers = buf.nioBuffers();
+            return writeBytesMultiple(in, nioBuffers, nioBuffers.length, readableBytes);
+        }
+    }
+
+    private boolean writeBytesMultiple(ChannelOutboundBuffer in, IovArray array) throws IOException {
+
+        long expectedWrittenBytes = array.size();
+        int cnt = array.count();
+
+        assert expectedWrittenBytes != 0;
+        assert cnt != 0;
+
+        boolean done = false;
+        long writtenBytes = 0;
+        int offset = 0;
+        int end = offset + cnt;
+        for (;;) {
+            long localWrittenBytes = Native.writevAddresses(fd, array.memoryAddress(offset), cnt);
+            if (localWrittenBytes == 0) {
+                // Returned EAGAIN need to set EPOLLOUT
+                setEpollOut();
+                break;
+            }
+            expectedWrittenBytes -= localWrittenBytes;
+            writtenBytes += localWrittenBytes;
```

Comment:
To avoid redundant computation would it make sense to compute this onces at the end?  For example:

```java
long initialExpectedWrittenBytes = expectedWrittenBytes;

for (;;) { /*do writes and decrement expectedWrittenBytes*/ }

in.removeBytes(initialExpectedWrittenBytes - expectedWrittenBytes);
```